### DISCUSSION TITLE
fvOptions for scalar source term

### DISCUSSION THREAD

**Discussion by uttamcadambi07** [mdolab/dafoam#650] [General]
Hi,

My goal is to add a point source of scalar in my domain and use ```runPrimal``` option in DAFoam to run the case (my ultimate goal is optimization but let us start with this simple case)

The test case is a very simple cylinder  flow with a circular cylinder placed in a domain at Reynolds number = 150. The incoming flow is U = 1m/s and we use a SpalartAllmaras turbulence model to get the mean flow fields. 

Now this is straightforward, everything converges well.

I want to now introduce a scalar source of constant strength to this domain. I want to introduce it along a constant line JUST after the inlet patch. 

I do it using  my ```controlDict``` file which goes like this

```
scalar3


		{
		
		type scalarTransport;
		libs ("libsolverFunctionObjects.so");
		enabled true;
		writeControl timeStep;
		writeInterval 500;
		log yes;
		nCorr 2;
		DT 0.0067;
		field T;
		fvOptions
		{
			scalar3
			{
				type	scalarFixedValueConstraint;
				active	true;
				
				scalarFixedValueConstraintCoeffs
				{
					timeStart		0;
					duration		100000;
					selectionMode	cellZone;
					cellZone		source1;
					
					fieldValues
						{
						T 1;
						}
				}
			}
		}
		
		}
```

Of course i run topoSet to set the cells that will hold this scalar source (which is seen as ```cellZone  source1```). 

When I run this case in OpenFOAM, I get something like this
![sheet](https://github.com/mdolab/dafoam/assets/10186475/db1802e7-c92d-4af8-9503-3255e3b89da7)

This is expected and physical - the scalar is swept away by the flow and diffuses throughout the domain following the fluid velocity. 

Now when I try to do the same thing with DAFoam, my solver exits in 1 iteration. 

The scalar residual is 0 (which is because i set the inlet patch to be fixedValue 0) BUT I have set the cellZone to contain the scalar strength of 5 units from t = 0. Which means the scalar field is not uniformly 0 at t = 0, it has a non-zero value along a thin line which is located just downstream of the inlet patch (if you look closely you can see it in the attached figure above). 

So this means that DAFoam is not able to pick up the source initialized through fvOptions. Why is this the case? Here is the source code from DASimpleTFoam 
```
{
    alphat = turbulencePtr_->nut() / Prt;
    alphat.correctBoundaryConditions();

    volScalarField alphaEff("alphaEff", turbulencePtr_->nu() / Pr + alphat);

    fvScalarMatrix TEqn(
        fvm::div(phi, T)
        - fvm::laplacian(alphaEff, T)
        - fTSource);

    TEqn.relax();

    // get the solver performance info such as initial
    // and final residuals
    SolverPerformance<scalar> solverT = TEqn.solve();

    this->primalResidualControl<scalar>(solverT, printToScreen, printInterval, "T");

    // bound T
    DAUtility::boundVar(allOptions, T, printToScreen);
}
```

is it because there is no fvOptions? 

So to circumvent the problem , I set the inlet patch to be 0.01 and the internalField of the scalar to be 0.01

This is to ensure that the residual at t = 1 is not zero and the solver does not exit. 

I get this

![sheet_1](https://github.com/mdolab/dafoam/assets/10186475/0da8d300-3646-4a63-9a01-20c3e05915f1)

Please can you tell me how to fix this issue?

**Comment by friedenhe** [comment]
Do you want to add this source term to the U equation or the T equation? If the former, you should add it as the first component of the fvSource term in UEqn. If the later, you need to add a fvSourceT term into TEqn in DASimpleTFoam. DAFoam should support fvOption, but the fvOption is not in TEqn of DASimpleTFoam.

**Reply by uttamcadambi07** [reply]
I want to add the source term to the T equation.

So I cannot add the fvSourceT term (because i want to use this as a design variable to optimize it). So my plan was to use fvOptions to add a scalar source term of fixed strength to a cellSet and then use DAFoam to optimize fvSourceT. But the issue is that DAFoam is not picking up the source added through fvOptions. 

I modified TEqnSimple.H to include the fvOptions in the TEqn as follows

```
{
    alphat = turbulencePtr_->nut() / Prt;
    alphat.correctBoundaryConditions();

    volScalarField alphaEff("alphaEff", turbulencePtr_->nu() / Pr + alphat);

    fvScalarMatrix TEqn(
        fvm::div(phi, T)
        - fvm::laplacian(alphaEff, T)
        - fTSource
        - fvOptions(T));

    TEqn.relax();

    // get the solver performance info such as initial
    // and final residuals
    SolverPerformance<scalar> solverT = TEqn.solve();

    this->primalResidualControl<scalar>(solverT, printToScreen, printInterval, "T");

    // bound T
    DAUtility::boundVar(allOptions, T, printToScreen);
}

```

here fTSource is the source term that I wish to optimize. So I would like to have two source terms - one that is of constant strength and the other that will be optimized. But even with this modification I am not able to detect the fixed source term (which should be represented by fvOptions). I checked openFoam ```scalarTransportFoam``` and found this

```
fvScalarMatrix TEqn
            (
                fvm::ddt(T)
              + fvm::div(phi, T)
              - fvm::laplacian(DT, T)
             ==
                fvOptions(T)
            );

```

since this works for OpenFOAM, I tried replicating the same in DAFoam

It is like the periodic hill case where if I want to optimize the source term  then I hardcode a "dp" force that ensures constant bulk Reynolds number but then optimize fvSource instead.

**Comment by friedenhe** [comment]
In this case, you can do a dirty workaround by adding another source term called fvSource1 into the T equation. Then, you can prescribe the driving source term for fvSource1.

**Comment by Nillman2021** [comment]
thank u bro, you have save my life.
I have already struggle with this for about one whole week.
